src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs

```diff
@@ -960,17 +960,60 @@ private int RunCore(TextWriter consoleOutput, ErrorLogger errorLogger, Cancellat
                     // We pass it to the generators, which will realize any symbols they require. 
                     compilation = RunGenerators(compilation, Arguments.ParseOptions, generators, analyzerConfigProvider, additionalTextFiles, diagnostics);
 
-                    var generatedSyntaxTrees = compilation.SyntaxTrees.Skip(Arguments.SourceFiles.Length);
-                    if (Arguments.AnalyzerConfigPaths.Length > 0)
+                    bool hasAnalyzerConfigs = !Arguments.AnalyzerConfigPaths.IsEmpty;
+                    bool hasGeneratedOutputPath = !string.IsNullOrWhiteSpace(Arguments.GeneratedFilesDirectory);
+
+                    var generatedSyntaxTrees = compilation.SyntaxTrees.Skip(Arguments.SourceFiles.Length).ToImmutableArray();
```

Comment:
Why `ToImmutableArray` vs. say `ToList`? This doesn't ever appear to be used as an `ImmutableArray<T>` hence seems like this is just adding some extra overhead.

Reply:
I needed the length; I am a cargo cult programmer so I used `ImmutableArray` because we use it everywhere else instead of lists.

More seriously, is there a perf preference between list and regular array? My intuition here is to use `.ToArray()` as we only need the length and to iterate over it.

Reply:
On framework `ToList` is generally faster than `ToArray` due to having one less allocation. On .NET Core it's somewhat the opposite because they use better builders (got schooled on that fix on twitter recently). Would probably go with `ToList` for now.

Reply:
Segmented arrays would be the best case scenario here. I've completed the implementation of `SegmentedArray<T>` but I'm still working through the details to get `SegmentedList<T>` in place.